pkg/registry/core/rest/storage_core.go

```diff
@@ -83,24 +89,91 @@ type GenericLegacyRESTStorageProvider struct {
+}
+
+type ServicesConfig struct {
+	// Service IP ranges
+	ClusterIPRange          net.IPNet
+	SecondaryClusterIPRange net.IPNet
+	NodePortRange           utilnet.PortRange
+
+	IPRepairInterval time.Duration
+}
+
+type rangeRegistries struct {
+	clusterIP          rangeallocation.RangeRegistry
+	secondaryClusterIP rangeallocation.RangeRegistry
+	nodePort           rangeallocation.RangeRegistry
+}
 
-	// ServiceIPRange is used to build cluster IPs for discovery.
-	ServiceIPRange net.IPNet
+type legacyProvider struct {
+	Config
+
+	primaryServiceClusterIPAllocator ipallocator.Interface
+	serviceClusterIPAllocators       map[api.IPFamily]ipallocator.Interface
+	serviceNodePortAllocator         *portallocator.PortAllocator
+
+	startServiceNodePortsRepair, startServiceClusterIPRepair func(onFirstSuccess func(), stopCh chan struct{})
+}
+
+func New(c Config) (*legacyProvider, error) {
+	rangeRegistries, serviceClusterIPAllocator, serviceIPAllocators, serviceNodePortAllocator, err := c.newServiceIPAllocators()
+	if err != nil {
+		return nil, err
+	}
+
+	p := &legacyProvider{
+		Config: c,
+
+		primaryServiceClusterIPAllocator: serviceClusterIPAllocator,
+		serviceClusterIPAllocators:       serviceIPAllocators,
+		serviceNodePortAllocator:         serviceNodePortAllocator,
+	}
+
+	// create service node port repair controller
+	client, err := kubernetes.NewForConfig(c.LoopbackClientConfig)
+	if err != nil {
+		return nil, err
+	}
+	p.startServiceNodePortsRepair = portallocatorcontroller.NewRepair(c.Services.IPRepairInterval, client.CoreV1(), client.EventsV1(), c.Services.NodePortRange, rangeRegistries.nodePort).RunUntil
+
+	// create service cluster ip repair controller
+	if !utilfeature.DefaultFeatureGate.Enabled(features.MultiCIDRServiceAllocator) {
+		p.startServiceClusterIPRepair = serviceipallocatorcontroller.NewRepair(
+			c.Services.IPRepairInterval,
+			client.CoreV1(),
+			client.EventsV1(),
+			&c.Services.ClusterIPRange,
+			rangeRegistries.clusterIP,
+			&c.Services.SecondaryClusterIPRange,
+			rangeRegistries.secondaryClusterIP,
+		).RunUntil
+	} else {
+		p.startServiceClusterIPRepair = serviceipallocatorcontroller.NewRepairIPAddress(
+			c.Services.IPRepairInterval,
+			client,
+			&c.Services.ClusterIPRange,
+			&c.Services.SecondaryClusterIPRange,
+			c.Informers.Core().V1().Services(),
+			c.Informers.Networking().V1alpha1().IPAddresses(),
+		).RunUntil
+	}
 
-	// allocates ips for secondary service cidr in dual  stack clusters
-	SecondaryServiceIPRange net.IPNet
-	ServiceNodePortRange    utilnet.PortRange
+	return p, nil
 }
 
-func (c GenericLegacyRESTStorageProvider) NewRESTStorage(apiResourceConfigSource serverstorage.APIResourceConfigSource, restOptionsGetter generic.RESTOptionsGetter) (genericapiserver.APIGroupInfo, error) {
+func (c *GenericConfig) NewRESTStorage(apiResourceConfigSource serverstorage.APIResourceConfigSource, restOptionsGetter generic.RESTOptionsGetter) (genericapiserver.APIGroupInfo, error) {
```

Comment:
this had a value receiver and now a pointer, is there any implication?

Reply:
same question for the other methods in this file

Reply:
We usually default to pointer. Only when there are side-effects that we don't want to leak, then we use non-pointer receivers. These constructors here don't have side-effects.